airbyte-cdk/python/airbyte_cdk/sources/declarative/migrations/legacy_to_per_partition_state_migration.py

```diff
@@ -0,0 +1,81 @@
+from airbyte_cdk.sources.declarative.models import DatetimeBasedCursor, SubstreamPartitionRouter
+
+
+def _is_already_migrated(stream_state: Mapping[str, Any]) -> bool:
+    return "states" in stream_state
+
+
+class LegacyToPerPartitionStateMigration(StateMigration):
+    """
+    Transforms the input state for per-partitioned streams from the legacy format to the low-code format.
+    The cursor field and partition ID fields are automatically extracted from the stream's DatetimebasedCursor and SubstreamPartitionRouter.
+
+    Example input state:
+    {
+    "13506132": {
+      "last_changed": "2022-12-27T08:34:39+00:00"
+    }
+    Example output state:
+    {
+      "partition": {"id": "13506132"},
+      "cursor": {"last_changed": "2022-12-27T08:34:39+00:00"}
+    }
+    """
+
+    def __init__(
+        self,
+        partition_router: SubstreamPartitionRouter,
+        cursor: DatetimeBasedCursor,
+        config: Mapping[str, Any],
+        parameters: Mapping[str, Any],
+    ):
+        self._partition_router = partition_router
+        self._cursor = cursor
+        self._config = config
+        self._parameters = parameters
+
+    def should_migrate(self, stream_state: Mapping[str, Any]) -> bool:
+        if _is_already_migrated(stream_state):
+            return False
+
+        # There is exactly one parent stream
+        number_of_parent_streams = len(self._partition_router.parent_stream_configs)
+        if number_of_parent_streams != 1:
+            # There should be exactly one parent stream
+            return False
+        """
+        The expected state format is
+        "<parent_key_id>" : {
+          "<cursor_field>" : "<cursor_value>"
+        }
+        """
+        if stream_state:
+            for key, value in stream_state.items():
+                if isinstance(value, dict):
+                    keys = list(value.keys())
+                    if len(keys) != 1:
+                        # The input partitioned state should only have one key
+                        return False
+                    if keys[0] != self._cursor_field():
+                        # Unexpected key. Found {keys[0]}. Expected {self._cursor.cursor_field}
+                        return False
+        return True
+
+    def migrate(self, stream_state: Mapping[str, Any]) -> Mapping[str, Any]:
+        partition_key_field = self._eval_partition_key()
+        states = [{"partition": {partition_key_field: key}, "cursor": value} for key, value in stream_state.items()]
+        return {"states": states}
+
+    def _cursor_field(self) -> str:
+        cursor_field = InterpolatedString.create(self._cursor.cursor_field, parameters=self._parameters)
+        return str(cursor_field.eval(self._config))
+
+    def _eval_partition_key(self) -> str:
+        partition_key = InterpolatedString.create(self._partition_router.parent_stream_configs[0].parent_key, parameters=self._parameters)
```

Comment:
This can be the part of the `__post_init__()` instead of creating and evaluating in place. I don't see any dynamic logic so far for this to be constantly evaluated inside the method, when we could have this checked once in __post_init__

Reply:
good catch! fixed